File: 1535-find-the-winner-of-an-array-game/1535-find-the-winner-of-an-array-game.cpp
class Solution {
public:
    int getWinner(vector<int>& arr, int k) {
        int mx = arr[0];
        queue<int> q;
        for(int i=1; i<arr.size(); i++){
            mx = max(mx, arr[i]);
            q.push(arr[i]);
        }
        int curr = arr[0], streak = 0;
        while(!q.empty()){
            int opponent = q.front(); q.pop();
            if(curr > opponent){
                q.push(opponent);
                streak++;
            }
            else{
                q.push(curr);
                curr = opponent;
                streak = 1;
            }
            if(streak == k or curr == mx) return curr;
        }
        return -1;
    }
};

File: 1535-find-the-winner-of-an-array-game/1535-find-the-winner-of-an-array-game.java
class Solution {
    public int getWinner(int[] arr, int k) {
        int mx = arr[0], curr = arr[0], streak = 0;
        for(int i=1; i<arr.length; i++) mx = Math.max(mx, arr[i]);
        for(int i=1; i<arr.length; i++){
            int opponent = arr[i];
            if(curr > opponent){
                streak++;
            }
            else{
                curr = opponent;
                streak=1;
            }
            if(streak==k || mx==curr) return curr;
        }
        return -1;
    }
}